Continue merging flash layer branch to v0.x.x (Part 2)

AndrejMitrovic opened the pull request:
Let's see if this works.

AndrejMitrovic committed:
Integrate the execution engine into Agora

This makes the Engine in charge of signature verification
as well as running the lock + unlock scripts.

Note: Temporarily set Faucet image to the one targeting
this flash-layer branch.

AndrejMitrovic committed:
Add script network tests

AndrejMitrovic committed:
Make toPushOpcode() non-unittest

It's useful for other routines.

AndrejMitrovic committed:
tests: add getRegistry()

It will be needed for non-agora nodes to
be able to connect to agora nodes in the
Flash layer tests.

AndrejMitrovic committed:
Add support for non-interactive threshold signatures

The `CHECK_MULTI_SIG` and `VERIFY_MULTI_SIG` enable
writing script locks which require a threshold
number of valid signatures for the provided public keys.

In the following text:
  N = number of sigs
  M = number of keys

It supports any combination of `N / M` where `N <= M`
and where the number of keys or signatures does not
exceed 5.

The limit is in place due to the relatively expensive
signature validation process, as each signature is
compared with every key until it either validates or
until the next key is about to be tested.

AndrejMitrovic committed:
Script: Remove unused OP.INVALID opcode

It didn't really serve any purpose except it
was taking space which could be used for
another opcode in the future.

AndrejMitrovic committed:
Script: Implement OP.PUSH_NUM* opcodes

These push single integer values between 0 and 5.
They will be mainly used by the upcoming CHECK_MULTI_SIG
opcode. They reduce the script size by not having to
push a [len:1, <num>] byte pair unlike in PUSH_BYTES_*.

AndrejMitrovic committed:
Fix send-tx client to use Schnorr

It was using EdDSA instead of Schnorr by mistake.

AndrejMitrovic committed:
Flash: Implement test-suite

Added a bunch of tests. Some of this code
needs to be folded into base classes later.

omerfirmak commented on source/agora/script/Opcodes.d:
```diff
@@ -61,67 +57,100 @@ public enum OP : ubyte
+    /// Supports non-interactive threshold multi-signatures. A combination
+    /// of `N / M` is supported where `N <= M` and where the number of keys
+    /// and signatures does not exceed 5.
+    /// Pops an item from the stock which is the count of public keys that
```
```suggestion
    /// Pops an item from the stack which is the count of public keys that
```

AndrejMitrovic commented:
I attached gdb after the timeout and got the stack traces:

[thread 1](https://gist.github.com/AndrejMitrovic/3024ad9f6dca05d201626cf0e1ed49d9)
[thread 2](https://gist.github.com/AndrejMitrovic/7907b366bd6204708de8580a993a2498)
[thread 3](https://gist.github.com/AndrejMitrovic/f23a7fd246dd226a1bb8693d8dfc0c57)
[thread 4](https://gist.github.com/AndrejMitrovic/f75f95671aa0d361b4eca97cfac0e50c)
[thread 5](https://gist.github.com/AndrejMitrovic/b9620340e808e92970cb6097d11e7e8e)
[thread 6](https://gist.github.com/AndrejMitrovic/9847d2a16377c8bf1f8849ecdf035982)
[thread 7](https://gist.github.com/AndrejMitrovic/ac2418f97fc390a129e7de7075496bd0)
[thread 8](https://gist.github.com/AndrejMitrovic/b5d5e9983d0f4346d0aefc020e815680)
[thread 9](https://gist.github.com/AndrejMitrovic/b27bfb502bdbf430409225a1e7be3bf6)

I don't know what to make of it yet.

Geod24 commented:
Which one segfaulted ?

AndrejMitrovic commented:
They didn't segfault, I attached GDB after a timeout. The unittests were running in the background.

AndrejMitrovic commented:
The issue is the tests never finish. And there's no logging either.

AndrejMitrovic commented:
Maybe I timed out too soon though. Could be that the tests just run a long time. I'll give this another try.

AndrejMitrovic commented:
> Maybe I timed out too soon though. Could be that the tests just run a long time. I'll give this another try.

So far it's been running for 40 minutes, way longer than the previous max at around ~20. I'll try to do another stack trace but after a longer timeout (maybe 30 mins).

AndrejMitrovic commented:
Running it for 30 minutes gets me a stack trace of only 3 threads:

[thread 1](https://gist.github.com/AndrejMitrovic/3a998eb06b6a59a97fbfd933433b1cd8)
[thread 2](https://gist.github.com/AndrejMitrovic/36312938190b20c50e98296a077d3e2c)
[thread 3](https://gist.github.com/AndrejMitrovic/020d3afe958632380ba3a305b6174f70)

AndrejMitrovic commented:
I'm trying to disable some tests individually now.

omerfirmak committed:
Flash: Implement Network module

This commit adds a graph representation of the lightning network and additional
book keeping needed to be done by the FlashNode.

It is missing removal of closed channels, does not take channel fees and balances
in to account.

AndrejMitrovic committed:
Flash: Implement Channel class

This class is responsible for maintaining a
single 2-party channel. There may be multiple
channels open between two parties, but the Channel
class only cares about a single channel instance.

The channel's ID is derived from the funding transaction's
UTXO hash. The channel becomes open once the funding
transaction is externalized in the blockchain. The channel
becomes closed once the closing transaction is externalized,
or may alternatively become closed after the externalization
of the trigger / update and settlement transactions.

AndrejMitrovic committed:
Flash: Implement Update / Settle signer

This `UpdateSigner` class is responsible for
safely signing & collecting settlement and update
transaction pairs.

The settlement multisig must be validated before
the update transaction signature is revealed,
otherwise the node risks having their funds locked
in the channel if this was the first (trigger)
transaction pair.

AndrejMitrovic committed:
Flash: Implement FlashAPI

Some of these may be replaced in the future as HTLC
support is added. Each API endpoint which is expected
to return a value may also return an error code with a
diagonstical message.

AndrejMitrovic committed:
Flash: Implement lock scripts

The `Scripts` module contains the ability to create
lock scripts for trigger / update transactions,
as well as key-derivation methods for signing &
validating the update & settlement transactions.

AndrejMitrovic committed:
Flash: Implement API types

These types are used in the API.

They contain the `ChannelState`, the various update / settlement
transaction types, the nonce types used for signing,
balance types for various balance requests and updates,
and a `Result` struct to have a consistent error-handling mechanism.

AndrejMitrovic committed:
Flash: Implement error codes

These are the various error codes which may be
returned by a call to a Flash API endpoint.

Some errors are recoverable, while others signal
that something went wrong. The errors typically
come with an extra diagnostical message.

AndrejMitrovic committed:
Flash: Implement ChannelConfig

The `ChannelConfig` contains all of the information necessary
to open a new channel with a peer.

It contains only public information, leaking this data cannot
cause a loss of funds - although it may cause privacy issues.

AndrejMitrovic committed:
Flash: Implement Invoice definition

The invoice is used to route HTLC payments
across the network. The node shares the hash
of a preimage to the party which intends to
pay for the invoice.

The destination node can redeem the payment
if the reveal the secret to the counter-party
of the channel through which the payment
was routed.

The secrets are then revealed in the reverse
path of the payment, thereby updating all
the balances and collecting fees.

AndrejMitrovic committed:
Flash: Implement encrypted onion packets

These packets ensure that each hop node in
the payment route only knows to decrypt
their own information and may only
route the payment to the next channel.

Note: Currently the size of the onion is not static.
This should be improved for privacy / censorship
reasons (See LN RFC spec).

AndrejMitrovic committed:
Flash: Implement basic routing info

Currently there are no routing algorithms.
Only the path definition is implemented.

AndrejMitrovic committed:
Flash: Implement FlashNode

This class takes care of most book-keeping.
A separate control class will be defined
which allows client code to explicitly
create or pay for invoices.

AndrejMitrovic committed:
Flash: Implement the Control API

This API can be used to control the flash node,
for example creating new invoices or paying
for existing ones.

omerfirmak commented:
I don't know if you guys were able to reproduce this on your local machines. Limiting the unittests to only 2 cores, similar to the CI, reproduced the issue on my machine.

AndrejMitrovic commented:
I didn't know the CI runs on only two cores.. can't we increase this? It doesn't resolve the underlying problem but it will at least unblock us.

AndrejMitrovic commented:
Ah did you mean just using `dsinglethreaded`?

omerfirmak commented:
No, I meant actually forcing tests to run on only 2 cores. I used `taskset` on linux.

AndrejMitrovic committed:
Flash: Clean up some documentation

AndrejMitrovic committed:
Flash: Remove dead code / fix outdated docs

AndrejMitrovic committed:
Update cur seq ID on accepted payment / update

AndrejMitrovic committed:
Increase delay of test

AndrejMitrovic committed:
Get rid of unnecessary delays

There is now a retry mechanism so it
should just cover it.

AndrejMitrovic committed:
Improve signing process

AndrejMitrovic committed:
Make beginCollaborativeClose() asynchronous

This unblocks the main thread so the tests
can time out fast.

AndrejMitrovic committed:
Temporarily disable settle tx test

AndrejMitrovic committed:
Flush stdout to console

omerfirmak committed:
Flash: Propagate payment failure down the path

If a node on the path fails to route the payment
the already created HTLCs should unwind.

AndrejMitrovic committed:
Flash: Add better printouts

Some of the keys are well-known,
so this improves debugging.

AndrejMitrovic committed:
Printouts: Add indexing

AndrejMitrovic committed:
Reduce waiting time

AndrejMitrovic committed:
Payment routing should be synchronous

AndrejMitrovic committed:
Properly support two-way payments

AndrejMitrovic committed:
Flash: Make createOnionPacket() a free function

Also document it.

AndrejMitrovic committed:
Change onion packets to be fixed-size

It's an alternative and simpler solution
compared to LN's recursive onion definition.

Fixes #1614

omerfirmak committed:
Flash: Dont propagate unknown payment hashes to be dropped

omerfirmak committed:
Use local seq_id in update fiber

This is safer in case someone updates the cur_seq_id before our
task starts

omerfirmak committed:
Flash: Route errors information back to the origin

When a payment attempt fails, origin of the payment shoould be
informed about the point and reason of the failure. So that it can
retry in the light of this information.

omerfirmak committed:
Add flash test sleeps back

Tests fail without them

AndrejMitrovic committed:
Make Payload and EncryptedPayload static size

This makes it easy to reject invalid packages
and reduces possibility of DDoS as all packets
must have an exact physical size.

AndrejMitrovic committed:
Flash: Optimize Flash for fast payments

This rewrites a lot of code to use only
one fiber per channel (+1 global fiber per node).

The onChannelOpen() now spawns a fiber to
avoid blocking the calling channel fiber.

The per-channel fiber makes sure to only run
one signing at a time. Any new payments / updates
will be queued.

The polling isCollectingSignatures() has been
replaced with a confirmChannelUpdate() push
mechanism. Counter-parties in the signing
process will await this confirmation before
proceeding with any new channel updates.

Added waitForUpdateIndex() to allow the tests
to efficiently wait for a new update index
instead of forcing Thread.sleep() calls, which
were now removed.

Fixes https://github.com/bpfkorea/agora/issues/1644

AndrejMitrovic committed:
Flash: Add verification of HTLCs

This compares the HTLC parameters with the
value set in the decrypted payload.

Also fixed the createOnionPacket() routine
as it was creating the wrong height locks
and forwarded amounts.

Fixes #1683

AndrejMitrovic committed:
Fix createOnionPacket() routine

It was creating the wrong height locks
and forwarded amounts.

Additionally the calling code didn't
set the right initial lock height.

Added a unittest to verify intention.

omerfirmak committed:
Flash: Errors should go back towards the origin

omerfirmak committed:
Flash: Replace erroring node with erroring channel id

omerfirmak committed:
Flash: Fix dropped HTLCs not being propagated

omerfirmak committed:
Flash: Re-enable failing payment attempt

AndrejMitrovic committed:
Flash: Move core functionality into Node

Previously it lived in the test package.

AndrejMitrovic committed:
Fix shadowing variable

AndrejMitrovic committed:
Flash: Introduce ThinFlashNode / FlashFullNode / FlashValidator

The ThinFlashNode class enables creating thin Flash nodes
which aren't themselves a FullNode / Validator.

This also enables FullNode / Validator to now directly
inherit FlashNode and define the missing routines
for blockchain monitoring & sending transactions.

These embed a FlashNode inside their definition
and they directly implement the Flash API.

The `in` parameter had to be removed since
vibe.d does not support it yet (neither const).

Part of: #1717

omerfirmak committed:
Onion: Allow shared secret to be obtained from decryption step

omerfirmak committed:
Flash: Keep track of shared secrets

Channels should keep track of the shared secret that they use to
decrypt a message. This is needed in order to obfuscate the error
messages that may be caused by that particular payment.

omerfirmak committed:
Dont pass references of scoped vars to fibers

omerfirmak committed:
Flash: Obfuscate OnionError in each hop

omerfirmak committed:
Onion: Allow shared secret to be obtained from encryption step

omerfirmak committed:
Flash: Save path and shared secrets for each payment attempt

When a error comes back to the origin node, it will need these
to deobfuscate the error packet.

omerfirmak committed:
Flash: Deobfuscate error packet in origin node

omerfirmak committed:
Flash: Delete DroppedHTLC struct

omerfirmak committed:
Flash: Save incoming payment errors

omerfirmak committed:
Flash: Cleanup after a successful payment

omerfirmak committed:
Flash: Re-try failed payment attempts

AndrejMitrovic committed:
NetworkManager: Add ability to retrieve address

This should ideally use the name registry,
perhaps an alternate Flash name registry.

Currently it just looks for the Flash peer
among the list of FullNode peers (not ideal).

AndrejMitrovic committed:
Make onAcceptedBlock non-final

It will be derived in the Flash-based classes.

AndrejMitrovic committed:
Rename prettify => flashPrettify

This avoids conflicts with the existing
prettify function.

omerfirmak committed:
Flash: Fix failing test

With retries this test is doomed to fail since there is no
alternative path. Retries are tested in another test anyways.

omerfirmak committed:
Flash: Adjustable channel fees

Flash nodes can now broadcast channel updates to update
their minimum fee requirements. They are not respected
by anyone just yet.

AndrejMitrovic commented:
It looks like the test-suite now passes! That means we can begin merging all the flash work into v0.x.x mainline.

However I need to fix up the commits so they're atomic. Some of the commits don't compile on their own.

AndrejMitrovic committed:
Flash: Change stdout to logging

During development we used stdout for
easier debugging. But merging this into
mainline would cause stdout pollution.

AndrejMitrovic commented:
@omerfirmak I added a commit changing all stdout printouts to log calls. Also added these:

```
scope (failure) factory.printLogs();
```

The printouts will no longer appear in real-time or be interleaved, but they will at least be grouped per node and it will avoid stdout spam for everyone else.

AndrejMitrovic commented:
@omerfirmak I'll give you the honors to merge.
